src/Service/Bulk/BulkCreateUsersService.php

```diff
@@ -0,0 +1,294 @@
+                throw new LineItemNotFoundException($exceptionMessage);
+            }
+
+            $this->generateSlugData($lineItems);
+            $notExistLineItemsArray = $this->lineItemIds
+                ? array_diff($lineItemIds, array_keys($this->lineItemSlugs))
+                : array_diff($lineItemSlugs, array_values($this->lineItemSlugs));
+        }
+
+        if (empty($lineItemIds) && empty($lineItemSlugs)) {
+            $this->getAllLineItemSlugs();
+        }
+
+        $userNameLastIndexes = $this->getLastUserAssignedToLineItems($this->lineItemSlugs);
+
+        $userGroupAssignCount = 0;
+        if ($groupPrefix) {
+            $loadBalancer = $this->ltiInstanceLoadBalancerFactory->__invoke('userGroupId');
+            $userGroupIds = $loadBalancer->generateGroupIds($groupPrefix);
+            $userGroupAssignCount = (int) ceil(
+                count($userPrefixes) * count($this->lineItemSlugs) * $batchSize / count($userGroupIds)
+            );
+        }
+
+        $slugTotalUsers = $this->createBulkUserAssignmentData(
+            $userNameLastIndexes,
+            $userGroupIds,
+            $userPrefixes,
+            $userGroupAssignCount,
+            $groupPrefix,
+            $batchSize
+        );
+
+        $message = $this->getOperationResultMessage($slugTotalUsers, $userPrefixes);
+
+        return new UserCreationResult($message, $notExistLineItemsArray);
+    }
+
+    public function getOperationResultMessage(array $slugTotalUsers, array $userPrefix): string
+    {
+        $message = '';
+        $userPrefixString = implode(',', $userPrefix);
+
+        foreach ($slugTotalUsers as $slug => $totalUsers) {
+            $message .= sprintf(
+                "%s users created for line item %s for user prefix %s \n",
+                $totalUsers,
+                $slug,
+                $userPrefixString
+            );
+        }
+
+        return $message;
+    }
+
+
+    private function generateSlugData(array $lineItems): void
+    {
+        $lineItemSlugs = [];
+        foreach ($lineItems as $lineItem) {
+            $lineItemSlugs[$lineItem->getId()] = $lineItem->getSlug();
+        }
+        $this->lineItemSlugs = $lineItemSlugs;
+    }
+
+    private function createBulkUserAssignmentData(
+        array $userNameLastIndexes,
+        array $userGroupIds,
+        array $userPrefix,
+        int $userGroupAssignCount,
+        ?string $groupPrefix,
+        int $batchSize
+    ): array {
+        $slugWiseTotalUsersArray = [];
+        $automateCsvPath = $this->generatedUsersFilePath . date('Y-m-d');
+
+        if (!$this->filesystem->exists($automateCsvPath)) {
+            $this->filesystem->mkdir($automateCsvPath);
+        }
+
+        foreach ($userPrefix as $prefix) {
+            $csvPath = sprintf('%s/%s', $automateCsvPath, $prefix);
+            if (!$this->filesystem->exists($csvPath)) {
+                $this->filesystem->mkdir($csvPath);
+            }
```

Comment:
One comment/proposal for both this 3 lines and lines 166-168:
please move it to a different method.